### PR TITLE
added unit tests for productcomparison module

### PR DESCRIPTION
**Pull request by Arunkumar0610**
added tests for  #1037 #1069 #1076 
 **SimplCommerce.Module.ProductComparison
Services - ComparingProductService
Controllers - ComparingProductController**

### CODE REVIEW

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+using SimplCommerce.Module.ProductComparison.Areas.ProductComparison.Controllers;
+using SimplCommerce.Module.ProductComparison.Areas.ProductComparison.ViewModels;
+using SimplCommerce.Module.ProductComparison.Models;
+using SimplCommerce.Module.ProductComparison.Services;
+using Xunit;
+namespace SimplCommerce.Module.ProductComparison.Tests.Controllers
+{
+    public class ComparingProductControllerTests
+    {
+        #region PositiveTestCases
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:23]: Please remove the region

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+                contentLocalizationServiceMock.Object,
+                workContextMock.Object
+            );
+            var user = new User()
+            {
+                Id = userId,
+                UserName = "TestUser"
+            };
+            workContextMock.Setup(repo => repo.GetCurrentUser()).ReturnsAsync(user);
+            // Act
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:54]:
```suggestion

            // Act
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+            var result = await controller.AddToComparison(model);
+            // Assert
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:56]:
```suggestion

            // Assert
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+                contentLocalizationServiceMock.Object,
+                workContextMock.Object
+            );
+            var user = new User()
+            {
+                Id = userId,
+                UserName = "TestUser"
+            };
+            workContextMock.Setup(repo => repo.GetCurrentUser()).ReturnsAsync(user);
+            // Act
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:94]:
```suggestion

            // Act
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+            var result = await controller.Remove(productId);
+            // Assert
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:96]:
```suggestion

            // Assert
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+            var controller = new ComparingProductController(
+                userManagerMock.Object,
+                repositoryMock.Object,
+                comparingProductServiceMock.Object,
+                productPricingServiceMock.Object,
+                mediaServiceMock.Object,
+                contentLocalizationServiceMock.Object,
+                workContextMock.Object
+            );
+            // Act
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:181]:
```suggestion

            // Act
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+            var result = await controller.Index();
+            // Assert
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:183]:
```suggestion

            // Assert
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+                contentLocalizationServiceMock.Object,
+                workContextMock.Object
+            );
+            var user = new User()
+            {
+                Id = userId,
+                UserName = "TestUser"
+            };
+            workContextMock.Setup(repo => repo.GetCurrentUser()).ReturnsAsync(user);
+            // Act
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:224]:
```suggestion

            // Act
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+            var result = await controller.AddToComparison(model);
+            // Assert
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:226]:
```suggestion

            // Assert
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+                contentLocalizationServiceMock.Object,
+                workContextMock.Object
+            );
+            var user = new User()
+            {
+                Id = userId,
+                UserName = "TestUser"
+            };
+            workContextMock.Setup(repo => repo.GetCurrentUser()).ReturnsAsync(user);
+            // Act
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:266]:
```suggestion

            // Act
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+            var result = await controller.Remove(productId);
+            // Assert
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:268]:
```suggestion

            // Assert
```

File: test/SimplCommerce.Module.ProductComparison.Tests/SimplCommerce.Module.ProductComparison.Tests.csproj [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+        </PackageReference>
+    </ItemGroup>
+
+    <ItemGroup>
+        <ProjectReference Include="..\..\src\Modules\SimplCommerce.Module.ProductComparison\SimplCommerce.Module.ProductComparison.csproj" />
+        <ProjectReference Include="..\..\src\SimplCommerce.Infrastructure\SimplCommerce.Infrastructure.csproj" />
+    </ItemGroup>
+
+    <ItemGroup>
+      <Folder Include="Controllers\" />
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/SimplCommerce.Module.ProductComparison.Tests.csproj:29]: Check if these are needed

File: test/SimplCommerce.Module.ProductComparison.Tests/Services/ComparingProductServiceTests.cs [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+using SimplCommerce.Infrastructure.Data;
+using SimplCommerce.Module.Catalog.Models;
+using SimplCommerce.Module.ProductComparison.Models;
+using SimplCommerce.Module.ProductComparison.Services;
+using Xunit;
+namespace SimplCommerce.Module.ProductComparison.Tests.Services
+{
+    public class ComparingProductServiceTests
+    {
+        #region PositiveTestCases
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Services/ComparingProductServiceTests.cs:14]: Remove the region

**Review by hishamco — CHANGES_REQUESTED** [review]

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+                UserName = "TestUser"
+            };
+            workContextMock.Setup(repo => repo.GetCurrentUser()).ReturnsAsync(user);
+            
+
+            // Act
+            var result = await controller.Remove(productId);
+            
+
+            // Assert
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:285]:
```suggestion

            // Assert
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+            );
+            var user = new User()
+            {
+                Id = userId,
+                UserName = "TestUser"
+            };
+            workContextMock.Setup(repo => repo.GetCurrentUser()).ReturnsAsync(user);
+            
+
+            // Act
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:281]:
```suggestion
            // Act
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+                workContextMock.Object
+            );
+            var user = new User()
+            {
+                Id = userId,
+                UserName = "TestUser"
+            };
+            workContextMock.Setup(repo => repo.GetCurrentUser()).ReturnsAsync(user);
+            
+
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:54]:
```suggestion
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+            // Act
+            var result = await controller.AddToComparison(model);
+            
+
+            // Assert
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:59]:
```suggestion
            // Assert
```

**Review by hishamco — APPROVED** [review]

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+            );
+            var user = new User()
+            {
+                Id = userId,
+                UserName = "TestUser"
+            };
+            workContextMock.Setup(repo => repo.GetCurrentUser()).ReturnsAsync(user);
+            
+
+            // Act
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:97]:
```suggestion
            // Act
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+                contentLocalizationServiceMock.Object,
+                workContextMock.Object
+            );
+            
+
+            // Act
+            var result = await controller.Index();
+            
+
+            // Assert
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:192]:
```suggestion
            // Assert
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+            );
+            var user = new User()
+            {
+                Id = userId,
+                UserName = "TestUser"
+            };
+            workContextMock.Setup(repo => repo.GetCurrentUser()).ReturnsAsync(user);
+            
+
+            // Act
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:233]:
```suggestion
            // Act
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 21]
```diff
@@ -0,0 +1,287 @@
...
+            var result = await controller.AddToComparison(model);
+            
+
+            // Assert
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:237]:
```suggestion
            // Assert
```

**Review by hishamco — CHANGES_REQUESTED** [review]

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 22]
```diff
@@ -0,0 +1,283 @@
...
+                Id = userId,
+                UserName = "TestUser"
+            };
+            workContextMock.Setup(repo => repo.GetCurrentUser()).ReturnsAsync(user);
+            
+            // Act
+            var result = await controller.Remove(productId);
+            
+
+            // Assert
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:100]:
```suggestion
            // Assert
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 23]
```diff
@@ -0,0 +1,283 @@
...
+                repositoryMock.Object,
+                comparingProductServiceMock.Object,
+                productPricingServiceMock.Object,
+                mediaServiceMock.Object,
+                contentLocalizationServiceMock.Object,
+                workContextMock.Object
+            );
+            
+
+            // Act
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:187]:
```suggestion
            // Act
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Services/ComparingProductServiceTests.cs [hunk 24]
```diff
@@ -0,0 +1,139 @@
...
+        {
+            // Arrange
+            var fromUserId = 1;
+            var toUserId = 2;
+            var repositoryMock = new Mock<IRepository<ComparingProduct>>();
+            var comparingProducts = new List<ComparingProduct>()
+                {
+                    new ComparingProduct { UserId = fromUserId, ProductId = 1001 },
+                    new ComparingProduct { UserId = fromUserId, ProductId = 1002 },
+                };
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Services/ComparingProductServiceTests.cs:42]:
```suggestion
            var comparingProducts = new List<ComparingProduct>()
            {
                new ComparingProduct { UserId = fromUserId, ProductId = 1001 },
                new ComparingProduct { UserId = fromUserId, ProductId = 1002 },
            };
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Services/ComparingProductServiceTests.cs [hunk 25]
```diff
@@ -0,0 +1,139 @@
...
+            {
+                ProductId = 1004,
+                Product=new Product()
+                {
+                    Name="TestProduct3"
+                },
+                CreatedOn = DateTime.Now,
+                UserId=1
+            },
+        };
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Services/ComparingProductServiceTests.cs:103]:
```suggestion
            var listComparingProducts = new List<ComparingProduct>()
            {
                new ComparingProduct()
                {
                    ProductId = 1001,
                    Product=new Product()
                    {
                        Name="TestProduct"
                    },
                    CreatedOn = DateTime.Now,
                    UserId=1
                },
                new ComparingProduct()
                {
                    ProductId = 1002,
                    Product=new Product()
                    {
                        Name="TestProduct1"
                    },
                    CreatedOn = DateTime.Now,
                    UserId=1
                },
                new ComparingProduct()
                {
                    ProductId = 1003,
                    Product=new Product()
                    {
                        Name="TestProduct2"
                    },
                    CreatedOn = DateTime.Now,
                    UserId=1
                },
                new ComparingProduct()
                {
                    ProductId = 1004,
                    Product=new Product()
                    {
                        Name="TestProduct3"
                    },
                    CreatedOn = DateTime.Now,
                    UserId=1
                },
            };
```

**Review by hishamco — CHANGES_REQUESTED** [review]

File: test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs [hunk 26]
```diff
@@ -0,0 +1,281 @@
...
+using SimplCommerce.Module.Catalog.Services;
+using SimplCommerce.Module.Core.Extensions;
+using SimplCommerce.Module.Core.Models;
+using SimplCommerce.Module.Core.Services;
+using SimplCommerce.Module.ProductComparison.Areas.ProductComparison.Controllers;
+using SimplCommerce.Module.ProductComparison.Areas.ProductComparison.ViewModels;
+using SimplCommerce.Module.ProductComparison.Models;
+using SimplCommerce.Module.ProductComparison.Services;
+using Xunit;
+namespace SimplCommerce.Module.ProductComparison.Tests.Controllers
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Controllers/ComparingProductControllerTests.cs:19]:
```suggestion

namespace SimplCommerce.Module.ProductComparison.Tests.Controllers
```

File: test/SimplCommerce.Module.ProductComparison.Tests/Services/ComparingProductServiceTests.cs [hunk 27]
```diff
@@ -0,0 +1,139 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Moq;
+using SimplCommerce.Infrastructure.Data;
+using SimplCommerce.Module.Catalog.Models;
+using SimplCommerce.Module.ProductComparison.Models;
+using SimplCommerce.Module.ProductComparison.Services;
+using Xunit;
+namespace SimplCommerce.Module.ProductComparison.Tests.Services
```
Comment (hishamco) [inline test/SimplCommerce.Module.ProductComparison.Tests/Services/ComparingProductServiceTests.cs:10]:
```suggestion

namespace SimplCommerce.Module.ProductComparison.Tests.Services
```

**Review by hishamco — CHANGES_REQUESTED** [review]

**Comment by Arunkumar0610** [comment]
@hishamco Yes I ran the tests at my end everything works as excepted

**Comment by hishamco** [comment]
Please resolve the conflict

**Comment by Arunkumar0610** [comment]
@hishamco Resolved the conflict

**Review by hishamco — APPROVED** [review]

**Pull request merged**